File: api/crs/codegen/lib/crs_solve_tril/crs_solve_tril.c
#include "crs_solve_tril.h"
#include "crs_solve_tril_types.h"

void crs_solve_tril(const CRS_Matrix *A, emxArray_real_T *b)
{
  int b_i;
  int i;
  int i1;
  int k;
  int n;
  n = A->row_ptr->size[0];
  for (i = 0; i <= n - 2; i++) {
    b_i = A->row_ptr->data[i];
    i1 = A->row_ptr->data[i + 1] - 2;
    for (k = b_i; k <= i1; k++) {
      b->data[i] -= A->val->data[k - 1] * b->data[A->col_ind->data[k - 1] - 1];
    }
    b->data[i] /= A->val->data[i1];
  }
}

void crs_solve_tril_initialize(void)
{
}

void crs_solve_tril_terminate(void)
{
}

File: api/crs/crs_solve_tril.m
function b = crs_solve_tril(A, b)
% crs_solve_tril Solves A\b, where A is lower triangular
%     b = crs_solve_tril(A, b)
%  The matrix is assumed to be lower triangular. The right-hand side vector
%  will be overwritten by A\b at output.

%#codegen -args {crs_matrix, coder.typeof(0, [inf,1])}

coder.inline('always');

n = int32(length(A.row_ptr) - 1);

for i = 1:n
    cend = A.row_ptr(i+1) - 1;
    assert(A.col_ind(cend) == i && A.val(cend) ~= 0);

    for k = A.row_ptr(i):(cend - 1)
        j = A.col_ind(k);
        b(i) = b(i) - A.val(k) * b(j);
    end

    b(i) = b(i) / A.val(cend);
end

function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     L = tril(rand(10,10)); b = rand(10,1);
%!     if condest(L) < 1.e5
%!         A = crs_createFromSparse(sparse(L));
%!         us = crs_solve_tril(A, b);
%!         assert(norm(L * us - b) <= 1.e-10);
%!     end
%! end
